### PR TITLE
Curriculum Learning --use-best-val-loss-from Argument

### PR DESCRIPTION
**Pull request by mmoffatt2**
Changes:
-run_experiment.py will now output csv files into a directory in csv_logs, instead of directly into csv_logs. This will help with the organization of different config runs as well as make curriculum learning easier.
-Added a new argument to run_experiment.py called "--use-best-val-loss-from". This takes the csv_logs directory and output directory of a previous config run, and will find and use the checkpoint with the lowest validation loss of that run to initialize the model.
-Added new arguments "--init_from" "prev_run" and "--prev_run_ckpt" which will be used by run_experiment.py to initialize the model based on the provided checkpoint.

Steps to test with config.json:
-Run "python3 run_experiments.py --config explorations/config.json --value_only --output_dir out_config"
&ensp;&ensp;-This should create a folder in "csv_logs" called "config-shakespeare_char" that contains all the csv files
-Run "python3 run_experiments.py --config explorations/random_seed_tests.json --value_only --output_dir out_random_seed_test --use-best-val-loss-from config-shakespeare_char out_config"

### CODE REVIEW

**Comment by gkielian** [comment]
> -run_experiment.py will now output csv files into a directory in csv_logs, instead of directly into csv_logs. This will help 
> with the organization of different config runs as well as make curriculum learning easier.
Awesome  :+1:  this is great, the folder was getting really crowded.

> -Added a new argument to run_experiment.py called "--use-best-val-loss-from". This takes the csv_logs directory and 
> output directory of a previous config run, and will find and use the checkpoint with the lowest validation loss of that run 
> to initialize the model.
Sounds good.

> -Added new arguments "--init_from" "prev_run" and "--prev_run_ckpt" which will be used by run_experiment.py to 
> initialize the model based on the provided checkpoint.
Interesting, sounds like a great plan.  So it starts with the --prev_run_ckpt then saves best val checkpoints to --out_dir then, is that correct?

> -Run "python3 run_experiments.py --config explorations/config.json --value_only --output_dir out_config"
> -This should create a folder in "csv_logs" called "config-shakespeare_char" that contains all the csv files
Interesting, so is the folder within csv_logs called named with "[config file prefix]-[name of dataset]"? 
Would there be a way to create arguments in run_experiments.py and train.py to directly set the name of the csv_logs subfolder?

> -Run "python3 run_experiments.py --config explorations/random_seed_tests.json --value_only --output_dir out_random_seed_test --use-best-val-loss-from config-shakespeare_char out_config"

Okay, sounds like a good starting point, next steps I'll try it out to take a closer look.

**Comment by mmoffatt2** [comment]
> > > -run_experiment.py will now output csv files into a directory in csv_logs, instead of directly into csv_logs. This will help
> > > with the organization of different config runs as well as make curriculum learning easier.
> > 
> > Awesome 👍 this is great, the folder was getting really crowded.
> > 
> > > -Added a new argument to run_experiment.py called "--use-best-val-loss-from". This takes the csv_logs directory and
> > > output directory of a previous config run, and will find and use the checkpoint with the lowest validation loss of that run
> > > to initialize the model.
> > 
> > Sounds good.
> > 
> > > -Added new arguments "--init_from" "prev_run" and "--prev_run_ckpt" which will be used by run_experiment.py to
> > > initialize the model based on the provided checkpoint.
> > 
> > Interesting, sounds like a great plan. So it starts with the --prev_run_ckpt then saves best val checkpoints to --out_dir then, is that correct?

Yes, starting iterations at 0 again.
> > 
> > > -Run "python3 run_experiments.py --config explorations/config.json --value_only --output_dir out_config"
> > > -This should create a folder in "csv_logs" called "config-shakespeare_char" that contains all the csv files
> > 
> > Interesting, so is the folder within csv_logs called named with "[config file prefix]-[name of dataset]"? Would there be a way to create arguments in run_experiments.py and train.py to directly set the name of the csv_logs subfolder?

Yes I should be able to add an argument for that. The current naming convention doesn't allow for repeat runs of the same config file without writing to the same folder, so that might work better. 
> > 
> > > -Run "python3 run_experiments.py --config explorations/random_seed_tests.json --value_only --output_dir out_random_seed_test --use-best-val-loss-from config-shakespeare_char out_config"
> > 
> > Okay, sounds like a good starting point, next steps I'll try it out to take a closer look.

**Comment by gkielian** [comment]
Okay, so after trying it it seems that the reason for the errors was the network architecture settings and context length were different in the random_seeds_test.json.

For tiktoken or using the same dataset for multiple training (e.g. if it nan's and we want to try more training with a different random seed from the last saved non-nan checkpoint) everything should work great.

For sentencepiece or character based tokenization, we'll probably need a common meta.pkl file but let's shelve that for a later.

**Comment by gkielian** [comment]
^To clarify above, in summary everything totally works :+1: 

Trying now with tiktoken across different datasets.

File: run_experiments.py [hunk 1]
```diff
@@ -11,8 +13,36 @@ def parse_args():
     parser.add_argument("--output_dir", type=str, default="out", help="Directory to place the set of output checkpoints.")
     parser.add_argument("--prefix", type=str, default='', help="Optional prefix for tensorboard_run_name and out_dir.")
     parser.add_argument("--value_only", action="store_true", help="Include only the values of the configuration parameters in the names.")
+    parser.add_argument("--use-best-val-loss-from", nargs=2, metavar=('csv_dir', 'output_dir'), type=str, default=['', ''],
+                        help="Grab the best val loss of the run given by the csv_dir. Then, use the corresponding ckpt from the matching output_dir")
     return parser.parse_args()
 
+def find_best_val_loss(csv_dir, output_dir):
+    csv_path = f"csv_logs/{csv_dir}"
```
Comment (gkielian) [inline run_experiments.py:21]: Let's keep the full path as the input for run_experiments.py, this way there will be tab completion available in the cli.

File: train.py [hunk 2]
```diff
@@ -219,6 +220,21 @@ def setup(self):
             self.model = GPT.from_pretrained(self.args.init_from, override_args)
             for k in ['n_layer', 'n_head', 'n_embd', 'block_size', 'bias', 'vocab_size']:
                 self.model_args[k] = getattr(self.model.config, k)
+        elif self.args.init_from == 'prev_run':
+            ckpt_path = os.path.join(self.args.prev_run_ckpt, 'ckpt.pt')
+            checkpoint = torch.load(ckpt_path, map_location=self.device)
+            checkpoint_model_args = checkpoint['model_args']
+            for k in ['n_layer', 'n_head', 'n_embd', 'block_size', 'bias', 'vocab_size']:
```
Comment (gkielian) [inline train.py:227]: Was thinking we can utilize something similar to this in the run_experiments.py to override the config settings for the next run from the last checkpoint's architecture settings (if --use-best-val-from-prior is active).

However, everything works right now, so lets queue it for another PR.

File: train.py [hunk 3]
```diff
@@ -315,8 +331,11 @@ def log_metrics(self, losses, lr, running_mfu, iter_num):
             self.write_to_csv(losses['train'].item(), losses['val'].item())
 
     def write_to_csv(self, *args):
-        os.makedirs(self.args.csv_dir, exist_ok=True)
-        csv_path = os.path.join(self.args.csv_dir, self.args.csv_name + ".csv")
+        csv_full_dir = self.args.csv_dir
+        if self.args.tensorboard_log:
+            csv_full_dir = f"{self.args.csv_dir}/{self.args.tensorboard_run_name.split('-')[0]}-{self.args.dataset}"
+        os.makedirs(csv_full_dir, exist_ok=True)
+        csv_path = os.path.join(csv_full_dir, self.args.csv_name + ".csv")
```
Comment (gkielian) [inline train.py:338]: Let's keep this intact and add an argparse arg to train.py to (optionally) set the name to a specified name in the csv_directory for this.

Can save this for a later PR though.

**Comment by gkielian** [comment]
Confirmed that it works with tiktoken tokenization.

I'll convert the feedback above into todos (github issues).
